DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md

```diff
@@ -0,0 +1,281 @@
+
+<img src="gzip_file_format.png" alt="gzip_file_format" width="400"/>
+
+The DEFLATE compression algorithm performs LZ77 Encoding followed by Huffman encoding (decompression decodes in the opposite order). For more information on the DEFLATE format and LZ77 and Huffman encoding/decoding, [start here](https://en.wikipedia.org/wiki/Deflate).
+
+A DEFLATE block is is structured as follows:
+- The first bit indicates whether this is the last block in a stream of blocks
+- The second and third bits indicate the block compression type
+  - 0x0 = **Uncompressed**
+  - 0x1 = **Statically compressed**
+  - 0x2 = **Dynamically compressed**
+  - 0x3 = reserved (N/A)
+- The remainder of the block depends on the block type:
+  - **Uncompressed blocks** contain some additional bits to realign the block to a byte boundary, followed by a 2-byte `length`, 2-byte `nlength` (where `length = ~nlength`), and finally the uncompressed data. The number of bytes to read is indicated by the 2-byte `length` that was read.
+  - **Statically compressed blocks** have a *static* Huffman table; a table defined in the DEFLATE format. Therefore the remainder of the data in the block is the payload to be decoded using this table.
+  - **Dynamically compressed blocks** are illustrated in the figure below. The second table is a list of code lengths for the literal and distance symbols. It is used to create the Huffman table for decoding the payload data. The first table is used to encode the second table. So, to decode a dynamically compressed block, the reader must read the first table, then decode the second table using the first table, create the Huffman table from the code lengths of the second table, and finally decode the payload using the Huffman table.
+
+The GZIP decompression engine in this reference design supports all three block types.
+
+<img src="deflate_dynamic_block_format.png" alt="deflate_dynamic_block_format" width="600"/>
+
+
+### GZIP Decompression FPGA Design
+The image below summarizes the full streaming GZIP decompression design. The orange kernels on the FPGA in the dashed box are the *actual* streaming GZIP decompression engine. The *Producer* and *Consumer* kernels are only used to stream data into and out of the GZIP decompression engine from device memory, respectively.
+
+<img src="gzip_decompression_engine_full.png" alt="gzip_decompression_engine_full" width="700"/>
+
+The figure below illustrates the GZIP decompression engine in isolation. It delineates the streaming GZIP decompressor from the streaming DEFLATE decompressor. The `GZIPMetaDataReader` kernel parses and strips away the GZIP header and footer data and forwards the consecutive DEFLATE blocks to the DEFLATE decompression engine.
```

Comment:
I think I made this comment on your slides as well, but I'd just add the DEFLATE dotted line box to the diagram above and skip this second diagram.  Your call though.